### PR TITLE
Feat: add rfc7807 rule

### PR DESCRIPTION
**Pull request by iosypov**
## What/Why/How?

Fix: #392 According the issue added new rules for check  rfc7807. https://datatracker.ietf.org/doc/html/rfc7807 

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+export const Operation4xxProblemDetailsRfc7807: Oas3Rule = () => {
+  return {
+    ResponsesMap(responses: Record<string, object>, { report, location }: UserContext) {
+      const codes = Object.keys(responses);
+      for (const code of codes) {
+        if (!/4[Xx0-9]{2}/.test(code)) continue;
+        const response = responses[code] as Oas3Response;
+        if (!response.content || !response.content['application/problem+json']) {
+          report({
+            message: 'Response `4xx` must have content type `application/problem+json`.',
```
Comment (adamaltman) [inline packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts:18]:
```suggestion
            message: 'Response `4xx` must have content-type `application/problem+json`.',
```

**Review by adamaltman — COMMENTED** [review]
Good start!

**Comment by tatomyr** [comment]
Please add this to all rules set: `core/src/config/all.ts`.

**Comment by iosypov** [comment]
> Please add this to all rules set: `core/src/config/all.ts`.

@tatomyr , done!

File: packages/core/src/rules/oas2/__tests__/operation-4xx-problem-details-rfc7807.test.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import { outdent } from 'outdent';
+import { lintDocument } from '../../../lint';
+import { parseYamlToDocument, replaceSourceWithRef, makeConfig } from '../../../../__tests__/utils';
+import { BaseResolver } from '../../../resolve';
+
+describe('Oas2 operation-4xx-problem-details-rfc7807', () => {
```
Comment (tatomyr) [inline packages/core/src/rules/oas2/__tests__/operation-4xx-problem-details-rfc7807.test.ts:6]: I'm not completely sure we have to introduce this check for **OAS 2** at all. The rule itself is about the `application/problem+json` content type's shape which was introduced in **OAS 3** I believe. 

@adamaltman ,  @RomanHotsiy , what's your opinion on this?

Comment (RomanHotsiy) [reply packages/core/src/rules/oas2/__tests__/operation-4xx-problem-details-rfc7807.test.ts:6]: OAS2 had `produces`/`consumes` fields. If this is hard to apply, let's implement it for oas3 only.

**Comment by adamaltman** [comment]
We also need:

- add file `docs/rules/operation-4xx-problem-details-rfc7807.md` following similar page template of `operation-4xx`
- add to docs/rules.md index file with link to details

I can help with that part if you need.

File: docs/rules/operation-4xx-problem-details-rfc7807.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+readable details of errors in a HTTP response to avoid the need to
+define new error response formats for HTTP APIs.
+
+Every operation with (400-499) HTTP response should have content-type `application/problem+json` and fields `title` and `type`. According to [specification](https://datatracker.ietf.org/doc/html/rfc7807).
+
+## Configuration
+
+| Option   | Type   | Description                                                                               |
+| -------- | ------ | ----------------------------------------------------------------------------------------- |
+| severity | string | Possible values: `off`, `warn`, `error`. Default `warn` (in `recommended` configuration). |
```
Comment (adamaltman) [inline docs/rules/operation-4xx-problem-details-rfc7807.md:23]: Hmmm, I don't think we should warn on this one in the recommended. I think it should be `off` in the recommended config.

Comment (iosypov) [reply docs/rules/operation-4xx-problem-details-rfc7807.md:23]: I will change this. Please, review all document

File: packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * Validation according rfc7807 - https://datatracker.ietf.org/doc/html/rfc7807
+ */
+export const Operation4xxProblemDetailsRfc7807: Oas3Rule = () => {
+  return {
+    ResponsesMap(responses: Record<string, object>, { report, location }: UserContext) {
+      const codes = Object.keys(responses);
+      for (const code of codes) {
+        if (!/4[Xx0-9]{2}/.test(code)) continue;
+        const response = responses[code] as Oas3Response;
+        if (!response.content || !response.content['application/problem+json']) {
```
Comment (RomanHotsiy) [inline packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts:16]: Response here can be a `$ref`. It's better to register on `Response` object and `skip` all non-4xx responses.

File: packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          report({
+            message: 'Response `4xx` must have content-type `application/problem+json`.',
+            location: location.child(code).key(),
+          });
+        }
+      }
+    },
+    Response: {
+      skip(response: Oas3Response) {
+        if (!response.content || !response.content['application/problem+json']) {
```
Comment (RomanHotsiy) [inline packages/core/src/rules/oas3/operation-4xx-problem-details-rfc7807.ts:26]: I think it's better to register on `MediaType` here.

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]

File: __tests__/lint-config/invalid-config--no-option/snapshot.js [hunk 6]
```diff
@@ -20,9 +20,24 @@ No configurations were defined in extends -- using built in recommended configur
...
+20 |   description: some desc
+21 |
+
+Error was generated by the operation-4xx-problem-details-rfc7807 rule.
+
+
 ../__fixtures__/valid-openapi.yaml: validated in <test>ms
 
-Woohoo! Your OpenAPI definition is valid. 🎉
+❌ Validation failed with 1 error.
```
Comment (RomanHotsiy) [inline __tests__/lint-config/invalid-config--no-option/snapshot.js:39]: why it breaks those tests?

File: packages/core/src/config/minimal.ts [hunk 7]
```diff
@@ -55,5 +56,6 @@ export default {
     'no-unused-components': 'warn',
     'no-undefined-server-variable': 'warn',
     'no-server-variables-empty-enum': 'error',
+    'operation-4xx-problem-details-rfc7807': 'off',
```
Comment (RomanHotsiy) [inline packages/core/src/config/minimal.ts:59]: seems like we don't need to add rules with 'off' here.
Looks like it doesn't work correctly and enables those rules in fact. Check the test snapshots updated above. They should not.

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]

File: docs/rules/operation-4xx-problem-details-rfc7807.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                type: string
+              title:
+                type: string
+```
+
+Example of **correct** operation response:
+
+```yaml
+post:
+  rresponses:
```
Comment (adamaltman) [inline docs/rules/operation-4xx-problem-details-rfc7807.md:64]:
```suggestion
  responses:
```

**Review by RomanHotsiy — APPROVED** [review]
Code looks good so I'm approving it.

It would be great to add at least one e2e test before merging.

File: docs/rules/operation-4xx-problem-details-rfc7807.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+| OAS | Compatibility |
+| --- | ------------- |
+| 2.0 | ❌            |
+| 3.0 | ✅            |
+| 3.1 | ✅            |
+
+## API design principles
+
+Problem Details for HTTP APIs as a way to carry machine-
```
Comment (adamaltman) [inline docs/rules/operation-4xx-problem-details-rfc7807.md:13]:
```suggestion
Problem Details for HTTP APIs are a way to carry machine-
```

File: docs/rules/operation-4xx-problem-details-rfc7807.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+readable details of errors in a HTTP response to avoid the need to
+define new error response formats for HTTP APIs.
+
+Every operation with (400-499) HTTP response should have content-type `application/problem+json` and fields `title` and `type`. According to [specification](https://datatracker.ietf.org/doc/html/rfc7807).
```
Comment (adamaltman) [inline docs/rules/operation-4xx-problem-details-rfc7807.md:17]:
```suggestion
Every operation with (400-499) HTTP response should have content-type `application/problem+json` and fields `title` and `type` according to the [specification](https://datatracker.ietf.org/doc/html/rfc7807).
```

**Review by adamaltman — APPROVED** [review]
I made very minor docs changes and now approving.

**Pull request merged**